externals/kyuubi-flink-sql-engine/src/main/java/org/apache/flink/client/deployment/application/executors/EmbeddedExecutorFactory.java

```diff
@@ -105,6 +114,24 @@ public boolean isCompatibleWith(final Configuration configuration) {
   public PipelineExecutor getExecutor(final Configuration configuration) {
     checkNotNull(configuration);
     Collection<JobID> executorJobIDs;
+    synchronized (bootstrapLock) {
```

Comment:
Emm... I'm not sure I get the idea.

If the code reaches here and `bootstrapJobIds == null` is true, another thread has no opportunity to initialize the static variables since they are waiting for the `bootstrapLock` to be released.

I suppose we should use a `CountDownLatch` here instead, to avoid blocking other threads to initialize the variables.

Reply:
IMH, if `bootstrapJobIds == null`, the current thread would release the lock and wait for `bootstrapLock.notifyAll()` to be called by the bootstrap thread.

Reply:
This could be verified by removing this line https://github.com/apache/kyuubi/blob/bf0307565cf1196447403c6a6f8e240ac01d1cd5/externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/FlinkSQLEngine.scala#L105 or add a delay like 30 seconds (which is the bootstrap timeout) to it.

Reply:
Sorry, my bad, the change makes sense to me

Reply:
No worries. I pushed a commit to improve the logs. I think it's good to merge when the CI turns green. Thanks!